File: backend_service/src/main/java/com/example/backend_service/model/Voucher.java
package com.example.backend_service.model;

import com.example.backend_service.common.DiscountType;
import com.example.backend_service.common.OwnerType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "vouchers")
@Data
@EqualsAndHashCode(callSuper = true)
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Voucher extends AbstractEntity<Long> {

    @Column(nullable = false, unique = true)
    private String code;

    @Column(name = "discount_value", nullable = false)
    private BigDecimal discountValue;

    @Enumerated(EnumType.STRING)
    @Column(name = "discount_type", nullable = false)
    private DiscountType discountType;

    @Enumerated(EnumType.STRING)
    @Column(name = "owner_type", nullable = false)
    private OwnerType ownerType;

    @Column(name = "shop_id")
    private Long shopId;

    @Column(name = "min_order_value")
    private BigDecimal minOrderValue;

    @Column(name = "max_discount")
    private BigDecimal maxDiscount;

    @Column(name = "start_date", nullable = false)
    private LocalDateTime startDate;

    @Column(name = "end_date", nullable = false)
    private LocalDateTime endDate;
}

File: fe/src/app/core/models/voucher.ts
export interface Voucher {
  id: number;
  code: string;
  discountValue: number;
  discountType: 'PERCENT' | 'FIXED';
  ownerType: 'SYSTEM' | 'SHOP';
  shopId: number | null;
  minOrderValue: number;
  maxDiscount: number;
  startDate: string;
  endDate: string;
  createdAt: string;
  updatedAt: string;
}

export interface VoucherRequest {
  code: string;
  discountValue: number;
  discountType: 'PERCENT' | 'FIXED';
  minOrderValue: number;
  maxDiscount: number;
  startDate: string;
  endDate: string;
}
